### DISCUSSION TITLE
Pasting from clipboard with content of three types `text/plain`+`text/html`+`text/x-gfm` causes duplication

### DISCUSSION THREAD

**Discussion by rybak** [community/community#65235] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Bug

### Context

Consider the following snippet of code (also available on [JSFiddle](https://jsfiddle.net/r3zsgha7/1/)), which puts several kinds of content into user's clipboard, when the user clicks on the button:

```html
<html>
<head><title>Demo</title></head>
<body>
	<script>
		function copy(e) {
			e.preventDefault();
			const plainText = "plain text: example.net";
			const html = 'HTML: <a href="https://example.net">example.net</a>';
			const gfm = "GFM: [example.net](https://example.net)";
			const handleCopyEvent = event => {
				event.stopPropagation();
				event.preventDefault();
				const clipboardData = event.clipboardData || window.clipboardData;
				clipboardData.setData('text/plain', plainText);
				clipboardData.setData('text/html', html);
				clipboardData.setData('text/x-gfm', gfm);
			};
			document.addEventListener('copy', handleCopyEvent);
			document.execCommand('copy');
			document.removeEventListener('copy', handleCopyEvent);
		}
	</script>
	<button onclick="copy(event);">Press me to copy</button>
</body>
</html>
```
This might not be the best code to do the interaction with clipboard from JavaScript, but it doesn't matter for the purposes of this bug report. I'm not very familiar with JavaScript development, so please bear with me.

### Steps to reproduce
1. Open [this JSFiddle](https://jsfiddle.net/r3zsgha7/1/) (or HTML code from above in a browser).
2. Click the button to put content into the clipboard;
3. Paste from clipboard into a field on GitHub, that has support for [GitHub Flavored Markdown](https://github.github.com/gfm/): PR, issue, discussion, comment under a commit, etc;
4. Observe the pasted text.

### Expected result
The target field contains the following Markdown:
```Markdown
GFM: [example.net](https://example.net)
```

### Actual result
The target field contain double the expected Markdown text, with a surprising conversion of plain text into GFM:
```Markdown
GFM: [example.net](https://example.net)plain text: [example.net](https://example.net/)
```

### Notes
- It requires pressing Undo (<kbd>Ctrl</kbd>+<kbd>Z</kbd>) twice to remove the pasted content.
- For GitHub's UI, I've tried combinations:

| formats | pasting result | notes |
| --- | --- | --- |
| `text/plain`+`text/x-gfm` | `GFM: [example.net](https://example.net)` | |
| `text/plain`+`text/html` | `plain text: [example.net](https://example.net/)` | with link conversion |
| `text/html`+`text/x-gfm` | `GFM: [example.net](https://example.net)` | |
|  `text/plain` | `plain text: example.net` | without conversion|
| `text/html` | N/A | nothing is pasted |
| `text/x-gfm` | `GFM: [example.net](https://example.net)` | |

- During investigation, I've tried using different URLs for demonstration purposes:
  ```js
  const plainText = "plain text: example.org";
  const html = 'HTML: <a href="https://example.net">example.net</a>';
  const gfm = "GFM: [example.com](https://example.com)";
  ```
  this somehow affected GitHub's logic and for `text/plain`+`text/html`+`text/x-gfm` the result is `GFM: [example.com](https://example.com)`.
- I've also checked how other editors behave for the triple `text/plain`+`text/html`+`text/x-gfm`:
  - On Bitbucket HTML is pasted (converted internally to Markdown). For `text/plain`+ `text/x-gfm`, the plain text is pasted.
  - GitLab's Markdown editor gives "Expected result". GitLab even has a [test for it](https://gitlab.com/gitlab-org/gitlab/-/blob/master/spec/frontend/content_editor/extensions/copy_paste_spec.js#L307-326).
  - poor man's HTML editor in a browser pasting result is the HTML content both for Firefox and Chromium.:
   ```html
   <html contenteditable=""><head><meta http-equiv="content-type" content="text/html; charset=UTF-8"></head><body></body></html>
   ```

**Comment by rybak** [comment]
> ### Actual result
>
> The target field contain double the expected Markdown text, with a surprising conversion of plain text into GFM:
>
> ```markdown
> GFM: [example.net](https://example.net)plain text: [example.net](https://example.net/)
> ```

After thinking about it a bit, it isn't that surprising. It's much easier to add Markdown to plain text than to [convert HTML](https://stackoverflow.com/a/1732454/1083697) to Markdown.

**Comment by rybak** [comment]
I published the [reproducer on JSFiddle](https://jsfiddle.net/r3zsgha7/1/) to make it easier to reproduce this bug.

**Comment by 0x6b** [comment]
In my scenario, `text/plain` + `text/html` causes the similar issue. With the following code ([JSFiddle](https://jsfiddle.net/pr5n03o7/)), 

```js
function copy(e) {
	e.preventDefault();
	const text = "[example.net](https://example.net)";
	const html = '<a href="https://example.net">example.net</a>';
	const handleCopyEvent = event => {
		event.stopPropagation();
		event.preventDefault();
		const clipboardData = event.clipboardData || window.clipboardData;
		clipboardData.setData('text/plain', text);
		clipboardData.setData('text/html', html);
	};
	document.addEventListener('copy', handleCopyEvent);
	document.execCommand('copy');
	document.removeEventListener('copy', handleCopyEvent);
}
```

The result is

```md
[[example.net](https://example.net/)](https://example.net)
```

Instead of just

```md
[example.net](https://example.net/)
```

**Reply by rybak** [reply]
Thank you. This kind of confirms the second row in the table in section "Notes" of the original post. But in your example, the weirdness of the logic for extracting a URL out of plain text (to turn the URL into Markdown syntax) is more prominent.

**Comment by alexanderadam** [comment]
This also leads to [confusion in other projects](https://github.com/0x6b/copy-selection-as-markdown/issues/126#issuecomment-2005382187).
